Some simple refactors & beginning of kotlin conversions of the player classes

Profpatsch opened the pull request:
#### What is it?
- [ ] Bugfix (user facing)
- [ ] Feature (user facing)
- [x] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR

First few commits are from https://github.com/TeamNewPipe/NewPipe/pull/11829 (merge first & rebase)
This partially conflicts with https://github.com/TeamNewPipe/NewPipe/pull/9592

#### APK testing
<!-- Use a new, meaningfully named branch. The name is used as a suffix for the app ID to allow installing and testing multiple versions of NewPipe, e.g. "commentfix", if your PR implements a bugfix for comments. (No names like "patch-0" and "feature-1".)  -->
<!-- Remove the following line if you directly link the APK created by the CI pipeline. Directly linking is preferred if you need to let users test.-->
The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR. You can find more info and a video demonstration [on this wiki page](https://github.com/TeamNewPipe/NewPipe/wiki/Download-APK-for-PR).

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

Isira-Seneviratne commented:
Might be a good idea to look into migrating to Media3: https://developer.android.com/media/media3/exoplayer

Profpatsch commented:
@Isira-Seneviratne that’s exactly what [newPlayer](https://github.com/TeamNewPipe/NewPlayer/) does. This is the initial work to integrate NewPlayer into NewPipe

Profpatsch commented:
i.e. adjust all the interfaces in a way that we can slot in NewPlayer easily into the existing player logic

Stypox commented on app/src/main/java/org/schabi/newpipe/player/Player.java:
```diff
@@ -579,7 +580,12 @@ private void destroyPlayer() {
+     */
+    public void saveAndShutdown() {
         if (DEBUG) {
             Log.d(TAG, "destroy() called");
```
```suggestion
            Log.d(TAG, "saveAndShutdown() called");
```

Stypox commented on app/src/main/java/org/schabi/newpipe/player/Player.java:
```diff
@@ -1970,6 +1976,9 @@ public void setFragmentListener(final PlayerServiceEventListener listener) {
 
+    /** Remove the listener, if it was set.
+     * @param listener listener to remove
+     * */
```
Why do you write Javadocs this way? In my Android Studio when I type `/**` I already get the correctly formatted Javadoc. It's not a big issue, but it's strange you are deviating from the style used throughout the project, especially if it's also the standard style in Android Studio

Profpatsch replied:
I think it’s because I always started typing after `/**`, and then Android Studio does not complete the arguments automatically. Since I started hitting Enter after `/**` the comment automatically gets created, you are right.

Stypox commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -0,0 +1,174 @@
+
+    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
+        if (DEBUG) {
+            Log.d(
+                TAG,
+                (
+                    "onStartCommand() called with: intent = [" + intent +
+                        "], flags = [" + flags + "], startId = [" + startId + "]"
+                    )
+            )
```
```suggestion
            Log.d(TAG, "onStartCommand() called with: intent = [$intent], flags = [$flags], startId = [$startId]")
```

Stypox commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -0,0 +1,174 @@
+        If the service is already started in foreground, requesting it to be started shouldn't
+        do anything
+         */
+        player.UIs().get<NotificationPlayerUi>(NotificationPlayerUi::class.java)
```
```suggestion
        player.UIs().get(NotificationPlayerUi::class.java)
```

Stypox commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -0,0 +1,174 @@
+        }
+
+        player.handleIntent(intent)
+        player.UIs().get<MediaSessionPlayerUi>(MediaSessionPlayerUi::class.java)
```
```suggestion
        player.UIs().get(MediaSessionPlayerUi::class.java)
```

Stypox commented on app/src/main/java/org/schabi/newpipe/player/ui/PlayerUiList.kt:
```diff
@@ -0,0 +1,124 @@
+import java.util.Optional
+
+class PlayerUiList(vararg initialPlayerUis: PlayerUi) {
+    var playerUis = GuardedByMutex(mutableListOf<PlayerUi>())
```
```suggestion
    private val playerUis = GuardedByMutex(mutableListOf<PlayerUi>())
```

Stypox commented on app/src/main/java/org/schabi/newpipe/player/ui/VideoPlayerUi.java:
```diff
@@ -761,7 +762,7 @@ public boolean isFullscreen() {
 
     /**
-     * Update the play/pause button ({@link R.id.playPauseButton}) to reflect the action
+     * Update the play/pause button (`R.id.playPauseButton`) to reflect the action
```
```suggestion
     * Update the play/pause button ({@code R.id.playPauseButton}) to reflect the action
```

Profpatsch replied:
Doesn’t link from comment this way

Stypox reviewed (changes requested):
Generally looks good to me, thanks! You can probably ignore my comments in `PlayerService` since it will be quite different after the Android Auto PR anyway.

snaik20 commented on app/src/main/java/org/schabi/newpipe/player/ui/VideoPlayerUi.java:
```diff
@@ -947,6 +948,7 @@ public void onShuffleClicked() {
         player.toggleShuffleModeEnabled();
     }
 
+    @SuppressLint("PrivateResource")
```
We shold probably stop relying on the exoplayer's private resources. 

Can you add a TODO here to replace the `exo_controls_repeat_all`, `exo_controls_repeat_one` and `exo_controls_repeat_off` drawables.

snaik20 reviewed (commented):
Just a few minor comments to look into

Profpatsch commented:
Rebased on top of https://github.com/TeamNewPipe/NewPipe/pull/11829

I removed the lateinit in `PlayerService`, because there is no reasonable way to make the player non-nullable now that the Android Auto change has landed.

Profpatsch commented on app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:
```diff
@@ -244,7 +244,7 @@ public void onServiceConnected(final PlayerService connectedPlayerService,
         hideSystemUiIfNeeded();
 
-        final Optional<MainPlayerUi> playerUi = player.UIs().get(MainPlayerUi.class);
+        final Optional<MainPlayerUi> playerUi = player.UIs().getOpt(MainPlayerUi.class);
```
nvm that’s exactly `get` now, so we can just migrate everything to it.

Stypox commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -148,203 +150,204 @@ public int onStartCommand(final Intent intent, final int flags, final int startI
-        if (Intent.ACTION_MEDIA_BUTTON.equals(intent.getAction())
-                && (player == null || player.getPlayQueue() == null)) {
+        if (Intent.ACTION_MEDIA_BUTTON == intent.action &&
+            (player == null || player!!.playQueue == null)
```
```suggestion
            (player?.playQueue == null)
```

Stypox commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -148,203 +150,204 @@ public int onStartCommand(final Intent intent, final int flags, final int startI
-            player.handleIntent(intent);
-            player.UIs().getOpt(MediaSessionPlayerUi.class)
-                    .ifPresent(ui -> ui.handleMediaButtonIntent(intent));
+        val p = player
+        if (p != null) {
```
player?.let { p ->

Stypox commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -148,203 +150,204 @@ public int onStartCommand(final Intent intent, final int flags, final int startI
         }
 
-        if (player != null && !player.exoPlayerIsNull()) {
+        if (player != null && !player!!.exoPlayerIsNull()) {
```
```suggestion
        if (player?.exoPlayerIsNull() == false) {
```

Stypox commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -148,203 +150,204 @@ public int onStartCommand(final Intent intent, final int flags, final int startI
             // We can't just pause the player here because it will make transition
             // from one stream to a new stream not smooth
-            player.smoothStopForImmediateReusing();
+            player!!.smoothStopForImmediateReusing()
```
```suggestion
            player?.smoothStopForImmediateReusing()
```

Stypox commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -148,203 +150,204 @@ public int onStartCommand(final Intent intent, final int flags, final int startI
-            return;
+    override fun onTaskRemoved(rootIntent: Intent?) {
+        super.onTaskRemoved(rootIntent)
+        if (player != null && !player!!.videoPlayerSelected()) {
```
```suggestion
        if (player?.videoPlayerSelected() == false) {
```

Stypox commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -148,203 +150,204 @@ public int onStartCommand(final Intent intent, final int flags, final int startI
-        mediaBrowserPlaybackPreparer.dispose();
-        mediaSession.release();
-        mediaBrowserImpl.dispose();
+        mediaBrowserPlaybackPreparer!!.dispose()
+        mediaSession!!.release()
+        mediaBrowserImpl!!.dispose()
```
```suggestion
        mediaBrowserPlaybackPreparer?.dispose()
        mediaSession?.release()
        mediaBrowserImpl?.dispose()
```

Stypox commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -148,203 +150,204 @@ public int onStartCommand(final Intent intent, final int flags, final int startI
             }
-            player.destroy();
-            player = null;
+            player!!.saveAndShutdown()
```
```suggestion
            player?.saveAndShutdown()
```

Stypox commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -148,203 +150,204 @@ public int onStartCommand(final Intent intent, final int flags, final int startI
-    private void cleanup() {
+    private fun cleanup() {
         if (player != null) {
             if (onPlayerStartedOrStopped != null) {
                 // notify that the player is being destroyed
```
```suggestion
            // notify that the player is being destroyed
            onPlayerStartedOrStopped?.accept(null)
```

Stypox commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -148,203 +150,204 @@ public int onStartCommand(final Intent intent, final int flags, final int startI
             if (onPlayerStartedOrStopped != null) {
                 // notify that the player is being destroyed
-                onPlayerStartedOrStopped.accept(null);
+                onPlayerStartedOrStopped!!.accept(null)
             }
```
```suggestion
```

Stypox commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -148,203 +150,204 @@ public int onStartCommand(final Intent intent, final int flags, final int startI
-    public static class LocalBinder extends Binder {
-        private final WeakReference<PlayerService> playerService;
+    class LocalBinder internal constructor(playerService: PlayerService) : Binder() {
+        private val playerService: WeakReference<PlayerService?>
```
```suggestion
        private val playerService: WeakReference<PlayerService>
```

Stypox commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -148,203 +150,204 @@ public int onStartCommand(final Intent intent, final int flags, final int startI
         }
     }
-    //endregion
 
+    //endregion
     //region Media browser
```
Why are all of these swapped?

Profpatsch replied:
Are they? It’s the previous region that ends and the new one that starts

Stypox replied:
Yeah and there's space in the middle, but anyway, this isn't important

Stypox commented on app/src/main/java/org/schabi/newpipe/player/ui/PlayerUiList.kt:
```diff
@@ -4,7 +4,7 @@ import org.schabi.newpipe.util.GuardedByMutex
 
 class PlayerUiList(vararg initialPlayerUis: PlayerUi) {
-    var playerUis = GuardedByMutex(mutableListOf<PlayerUi>())
+    private var playerUis = GuardedByMutex(mutableListOf<PlayerUi>())
```
Do this and remove the init (and move the init comment according to Kotlin's rules for constructor comments)
```suggestion
    private var playerUis = GuardedByMutex(mutableListOf<PlayerUi>(initialPlayerUis))
```

Stypox replied:
Can this be a `val`?

Stypox reviewed (changes requested):
After you fix the artifacts from the conversion to Kotlin feel free to merge this (or ask me for approval). Thanks!

snaik20 commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
Yupp, Can convert all Consumer and BiConsumer to lambdas

snaik20 commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -0,0 +1,352 @@
+     * by the [Consumer] can be null to indicate that the player is stopping.
+     * @param listener the listener to set or unset
+     */
+    fun setPlayerListener(listener: Consumer<Player?>?) {
```
You can use lambda `((Player?) -> Unit)` instead of using Consumer object. For Java, the equivalent would be `Function1<Player, Unit>`

Profpatsch replied:
Will be fixed later in the commit chain

snaik20 commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -0,0 +1,352 @@
+            // no one already and starting the service in foreground should not create any issues.
+            // If the service is already started in foreground, requesting it to be started
+            // shouldn't do anything.
+            player!!.UIs().get(NotificationPlayerUi::class.java)
```
Can be chained with optional unwrapping operator `?.`

snaik20 commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -0,0 +1,352 @@
+        rootHints: Bundle?
+    ): BrowserRoot {
+        // TODO check if the accessing package has permission to view data
+        return mediaBrowserImpl!!.onGetRoot(clientPackageName, clientUid, rootHints)
```
The `onGetRoot` method returns `@Nullable BrowserRoot`, so there’s no need to use a force unwrap here.

The default Java-to-Kotlin conversion in Android Studio isn’t very smart and doesn’t always identify such cases.

snaik20 commented on app/src/main/java/org/schabi/newpipe/player/PlayerService.kt:
```diff
@@ -0,0 +1,352 @@
+    }
+
+    class LocalBinder internal constructor(playerService: PlayerService) : Binder() {
+        private val playerService: WeakReference<PlayerService?>
```
You can combine declaration and assignment.
```suggestion
        private val playerService: WeakReference<PlayerService> = WeakReference<PlayerService>(playerService)
```

Profpatsch commented:
@snaik20 btw if you have some nitpicks/trivial improvements, it’s fine if you add a commit and push that directly on the branch; I’m guessing you have the editor open anyway and it prevents a bunch of back-and-forth
